src/models/Event.js

```diff
@@ -1,124 +1,86 @@
-
-    if (! events || ! events.length) {
-      return new HttpError('Failed to load event data.');
-    }
-
-    const formattedEvents = events.map((event) => ({
-      title: {
-        'en-US': event['Event Title (US-EN)'],
-        'es-MX': event['Event Title (ES-MX)'],
-      },
-      isPublished: event.Published,
-      date: event.Date,
-      startTime: event['Start Time'],
-      endTime: event['End Time'],
-      timezone: event.Timezone,
-      publicAddress: event['Public Address'],
-      city: event.City,
-      state: event.State,
-      zipcode: event.Zipcode,
-      latitude: event.Latitude,
-      longitude: event.Longitude,
-      rsvpLink: event['RSVP Link'],
-      rsvpCtaOverride: event['RSVP CTA'],
-    }));
-
-    const publishedEvents = formattedEvents.filter(({ isPublished }) => isPublished);
-
-    // TODO: We should make this smarter.
-    const openEvents = publishedEvents.filter(({ date }) => eventTime(date, true) > Date.now());
-
-    return openEvents;
+/**
+ * Represents an event, a la Mobilize America or a BSD event.
+ *
+ * Pulls events from mongodb, where an example document is like:
+ *   {
+ *     loc: { type: 'Point', coordinates: [-93.8549133, 41.6157869] },
+ *     title: {
+ *       'en-US': 'Waukee for Warren Coffee Hours',
+ *       'es-MX': 'Waukee for Warren Coffee Hours',
+ *     },
+ *     published: true,
+ *     date: new Date(1556114400),
+ *     startTime: new Date(1556114400),  // April 24, 2019 10:00 AM EDT
+ *     endTime: new Date(1556125200),
+ *     timezone: 'America/Chicago',
+ *     publicAddress: '1025 E Hickman Rd, Waukee IA 50263',
+ *     city: 'Waukee',
+ *     state: 'IA',
+ *     zipcode: '50263',
+ *     rsvpLink: 'https://events.elizabethwarren.com/event/88773/',
+ *     rsvpCtaOverride: null,
+ * }
+ *
+ * Returns list of event objects in that format.
+ */
+
+module.exports = (db) => {
+  const collection = db.collection('events');
+  const searchRadius = 300 * 1609;  // 300 miles in meters
+
+  async function _init() {
+    await collection.createIndex([ { startTime: 1 }, { highPriority: -1 }]);
+    await collection.createIndex({ mobilizeId: 1 });
+    await collection.createIndex([{ loc: '2dsphere' }, { highPriority: -1 }]);
   }
 
-  const loadEvents = asyncWrap(_loadEvents);
+  const init = asyncWrap(_init);
 
   /**
-   * Get a list of events that haven't hapended yet,
-   * sorted by how close they are to happening.
+   * Get a list of events that haven't happened yet,
+   * ordered by how soon they will happen.
    *
    * @return      {Array<Object>}
    */
   async function _getUpcomingEvents() {
-    const events = await loadEvents();
-
-    if (events instanceof HttpError) {
-      return events;
-    }
-
-    const sorted = events.sort(({ date: a }, { date: b }) => (
-      eventTime(a) - eventTime(b)
-    ));
-
-    return sorted;
+    const eventsCursor = await collection.find({
+      startTime: {
+        $gte : new Date(),
```

Comment:
A fun problem we're going to have to take a closer look at (which the prior version hacked around very badly) is how we deal with timezone differences between the client & the server (& the events in the DB)

Reply:
Not a blocker for merging this, just a heads-up

Reply:
Yeah great point.